src/test/java/org/apache/maven/plugins/shade/DefaultShaderTest.java

```diff
@@ -113,16 +83,67 @@ public Logger getChildLogger( final String s )
+        final String failureDebugMessage = logs.debugMessages.toString();
+        assertTrue(failureDebugMessage, logs.debugMessages.contains(
                 "We have a duplicate META-INF/MANIFEST.MF in src/test/jars/plexus-utils-1.4.1.jar" ));
     }
 
+    public void testOverlappingResourcesAreLoggedExceptATransformerHandlesIt() throws Exception {
+        final TemporaryFolder temporaryFolder = new TemporaryFolder();
+        final Set<File> set = new LinkedHashSet<>();
+        temporaryFolder.create();
+        final File j1 = temporaryFolder.newFile("j1.jar");
+        try (final JarOutputStream jos = new JarOutputStream(new FileOutputStream( j1 ) ) )
+        {
+            jos.putNextEntry(new JarEntry( "foo.txt" ));
+            jos.write("c1".getBytes(StandardCharsets.UTF_8));
+            jos.closeEntry();
+        }
+        final File j2 = temporaryFolder.newFile("j2.jar");
+        try (final JarOutputStream jos = new JarOutputStream(new FileOutputStream( j2 ) ) )
+        {
+            jos.putNextEntry(new JarEntry( "foo.txt" ));
+            jos.write("c2".getBytes(StandardCharsets.UTF_8));
+            jos.closeEntry();
+        }
+        set.add( j1 );
+        set.add( j2 );
+
+        final AppendingTransformer transformer = new AppendingTransformer();
+        final Field resource = AppendingTransformer.class.getDeclaredField( "resource" );
+        resource.setAccessible( true );
+        resource.set( transformer, "foo.txt" );
+
+        final ShadeRequest shadeRequest = new ShadeRequest();
+        shadeRequest.setJars( set );
+        shadeRequest.setRelocators( Collections.<Relocator>emptyList() );
+        shadeRequest.setResourceTransformers( Collections.<ResourceTransformer>singletonList( transformer) );
+        shadeRequest.setFilters( Collections.<Filter>emptyList() );
+        shadeRequest.setUberJar( new File( "target/foo-custom_testOverlappingResourcesAreLogged.jar" ) );
+
+        final DefaultShader shaderWithTransformer = new DefaultShader();
+        final MockLogger logWithTransformer = new MockLogger();
+        shaderWithTransformer.enableLogging( logWithTransformer );
+        shaderWithTransformer.shade( shadeRequest );
+
+        final DefaultShader shaderWithoutTransformer = new DefaultShader();
+        final MockLogger logWithoutTransformer = new MockLogger();
+        shaderWithoutTransformer.enableLogging( logWithoutTransformer );
+        shadeRequest.setResourceTransformers( Collections.<ResourceTransformer>emptyList() );
+        shaderWithoutTransformer.shade( shadeRequest );
+
+        temporaryFolder.delete();
+
+        assertTrue(logWithTransformer.warnMessages.toString(), logWithTransformer.warnMessages.isEmpty());
+        assertTrue(logWithoutTransformer.warnMessages.toString(), logWithoutTransformer.warnMessages.containsAll(
+               Arrays.<String>asList( "j1.jar, j2.jar define 1 overlapping resources:", "- foo.txt" ) ) );
```

Comment:
resources --> resource when it's singular

Reply:
fixed